File: Ahorn/entities/colorSwitch.jl
module VortexHelperColorSwitch
using ..Ahorn, Maple

const colors = ["blue", "rose", "orange", "lime"]

@mapdef Entity "VortexHelper/ColorSwitch" ColorSwitch(
			x::Integer, y::Integer,
			width::Integer=16, height::Integer=16,
			blue::Bool=false, rose::Bool=false, orange::Bool=false, lime::Bool=false, random::Bool=false)
			
const placements = Ahorn.PlacementDict(
    "Color Switch (All, Cycle) (Vortex Helper)" => Ahorn.EntityPlacement(
        ColorSwitch,
		"rectangle",
		Dict{String, Any}(
            "blue" => true, "rose" => true, "orange" => true, "lime" => true
        )
	),
    "Color Switch (All, Random) (Vortex Helper)" => Ahorn.EntityPlacement(
        ColorSwitch,
		"rectangle",
		Dict{String, Any}(
            "blue" => true, "rose" => true, "orange" => true, "lime" => true, "random" => true
        )
	)
)

const block = "objects/VortexHelper/onoff/switch"
const tileTint = Ahorn.colorTupleType((0.5, 0.5, 0.5, 1.0))
const backColor = (40, 40, 40, 255) ./ 255

Ahorn.minimumSize(entity::ColorSwitch) = 16, 16
Ahorn.resizable(entity::ColorSwitch) = true, true

Ahorn.selection(entity::ColorSwitch) = Ahorn.getEntityRectangle(entity)

function renderColorSwitch(ctx::Ahorn.Cairo.CairoContext, x::Number, y::Number, width::Number, height::Number)
	tilesWidth = div(width, 8)
    tilesHeight = div(height, 8)
	
	Ahorn.drawRectangle(ctx, x + 2, y + 2, width - 4, height - 4, backColor)

    for i in 2:tilesWidth - 1
        Ahorn.drawImage(ctx, block, x + (i - 1) * 8, y, 8, 0, 8, 8, tint=tileTint)
        Ahorn.drawImage(ctx, block, x + (i - 1) * 8, y + height - 8, 8, 16, 8, 8, tint=tileTint)
    end

    for i in 2:tilesHeight - 1
        Ahorn.drawImage(ctx, block, x, y + (i - 1) * 8, 0, 8, 8, 8, tint=tileTint)
        Ahorn.drawImage(ctx, block, x + width - 8, y + (i - 1) * 8, 16, 8, 8, 8, tint=tileTint)
    end

    Ahorn.drawImage(ctx, block, x, y, 0, 0, 8, 8, tint=tileTint)
    Ahorn.drawImage(ctx, block, x + width - 8, y, 16, 0, 8, 8, tint=tileTint)
    Ahorn.drawImage(ctx, block, x, y + height - 8, 0, 16, 8, 8, tint=tileTint)
    Ahorn.drawImage(ctx, block, x + width - 8, y + height - 8, 16, 16, 8, 8, tint=tileTint)
end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::ColorSwitch, room::Maple.Room)
	x = Int(get(entity.data, "x", 0))
    y = Int(get(entity.data, "y", 0))

    width = Int(get(entity.data, "width", 16))
    height = Int(get(entity.data, "height", 16))

    renderColorSwitch(ctx, x, y, width, height)
end

end

File: Loenn/entities/color_switch.lua
local drawableRectangle = require("structs.drawable_rectangle")
local drawableNinePatch = require("structs.drawable_nine_patch")
local vortexHelper = require("mods").requireFromPlugin("libraries.vortex_helper")

local colorSwitch = {}

colorSwitch.name = "VortexHelper/ColorSwitch"
colorSwitch.resizable = {true, true}
colorSwitch.minimumSize = {16, 16}

colorSwitch.placements = {
    {
        name = "all_cycle",
        data = {
            width = 16,
            height = 16,
            blue = true,
            rose = true,
            orange = true,
            lime = true,
            random = false,
            spriteDir = ""
        }
    },
    {
        name = "all_random",
        data = {
            width = 16,
            height = 16,
            blue = true,
            rose = true,
            orange = true,
            lime = true,
            random = true,
            spriteDir = ""
        }
    }
}

function colorSwitch.sprite(room, entity)
    local x, y = entity.x or 0, entity.y or 0
    local width, height = entity.width or 16, entity.height or 16

    local colors = vortexHelper.colorSwitchRoomColors(room)
    local frame = (entity.spriteDir or "") ~= "" and (entity.spriteDir .. "/switch") or "objects/VortexHelper/onoff/switch"
    local nine_patch_options = {
        mode = "border",
        borderMode = "repeat",
        color = colors.edgeColor
    }

    return {
        drawableRectangle.fromRectangle("fill", x + 1, y + 1, width - 2, height - 2, colors.bgColor):getDrawableSprite(),
        drawableNinePatch.fromTexture(frame, nine_patch_options, x, y, width, height)
    }
end

return colorSwitch
